test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp

```diff
@@ -1,54 +1,133 @@
-#include <stan/math/prim/mat.hpp>
+#ifndef TEST_UNIT_MATH_PRIM_MAT_PROB_VECTOR_RNG_TEST_HELPER_HPP
+#define TEST_UNIT_MATH_PRIM_MAT_PROB_VECTOR_RNG_TEST_HELPER_HPP
+
 #include <gtest/gtest.h>
 #include <boost/math/distributions.hpp>
 #include <boost/random/mersenne_twister.hpp>
+#include <stan/math/prim/mat.hpp>
 #include <test/unit/math/prim/scal/prob/util.hpp>
+#include <vector>
+
+/*
+ * Objects of this class can be used to generate sequences of non-random
+ * numbers.
+ */
+class non_random_generator {
+private:
+  int last_used_idx_;
 
-enum class Constraint { None, Positive, NonNegative };
+  const std::vector<double> real_values_ =
+    { -5.0, -4.0, -3.5, -1.0, -0.5, 0.0, 1.5, 2.8, 3.0, 5.0, 7.0 };
+
+public:
+  non_random_generator() : last_used_idx_(0) {}
+
+  size_t size() const {
+    return real_values_.size();
+  }
+
+  /*
+   * Get next value from non-random generator. Increment internal index to the
+   * next valid value (wrapping around to the beginning).
+   *
+   * @return A non-random number
+   */
+  double operator()() {
+    last_used_idx_ = (last_used_idx_ + 1) % real_values_.size();
+    return real_values_[last_used_idx_];
+  }
+};
 
 /*
- * Resize v to be length N and fill v with values that satisfy the given constraint c.
+ * Build a length N vector-like variable of type T1. Fill it with numbers
+ * generated from the callable non_random_generator_instance that satisfy the
+ * constraint check.
+ *
+ * This is a functor instead of a function because it takes advantage of partial
+ * template specialization.
  *
  * @tparam T1 Type of v
- * @param v Vector to fill with values
- * @param c Constraint that values must fulfill
+ * @tparam T_check Callable type that takes in one double and returns a bool
+ * @tparam T_generator Callable type that takes no arguments and returns a
+ * double
  * @param N Number of elements that v will be resized to
+ * @param check Callable for validating p
+ * @param non_random_generator_instance Callable for generating possible
+ * parameters
+ * @throws std::domain_error if non_random_generator_instance cannot satisfy
+ * check constraint
  */
-template <typename T1>
-void generateValues(T1& v, Constraint c, int N) {
-  v.resize(N);
-  
-  if (c == Constraint::None) {
-    for (int i = 0; i < N; i++)
-      v[i] = i - 1;
-  } else if (c == Constraint::Positive) {
-    for (int i = 0; i < N; i++)
-      v[i] = i + 1;
-  } else if (c == Constraint::NonNegative) {
-    for (int i = 0; i < N; i++)
-      v[i] = i;
+template<typename T1>
+struct generate_values {
+  template <typename T_check, typename T_generator>
+  T1 operator()(int N, T_check check,
+                T_generator &non_random_generator_instance) {
+    T1 vec(N);
+
+    for (int i = 0; i < vec.size(); i++) {
+      double v;
+
+      // There are only .size() possible non_random_numbers to try
+      for (int j = 0; j < non_random_generator_instance.size(); j++) {
```

Comment:
this makes it seem pretty clear that the concept of a `non_random_generator_instance` is just a list, no?  :P

Reply:
Haha, yeah, but it saves its state in a weird way. A bunch of things use it, but they just start from where the last instance left off (this is where it's like the random number generator -- things really don't have a clue what they'll see next).

It's unfortunate those generate_values functions aren't member functions. The way I'm using them (to return either a std::vector or a double) I'd need partial function template specialization. Unless the value is returned through an input argument, bleh.